src/ucp/core/ucp_ep.c

```diff
@@ -781,29 +781,37 @@ void ucp_ep_cleanup_lanes(ucp_ep_h ep)
 
     for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
         uct_ep = ep->uct_eps[lane];
-        if (uct_ep != NULL) {
-            ucs_debug("ep %p: purge uct_ep[%d]=%p", ep, lane, uct_ep);
-            uct_ep_pending_purge(uct_ep, ucp_destroyed_ep_pending_purge, ep);
-        }
-    }
 
-    for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
-        uct_ep = ep->uct_eps[lane];
+        /* set UCT EP to failed UCT EP to make sure if UCP EP won't be destroyed
+         * due to some UCT EP discarding procedures are in-progress and UCP EP
+         * may get some operation completions which could try to dereference
+         * its lanes */
+        ep->uct_eps[lane] = &ucp_failed_tl_ep;
```

Comment:
BTW, maybe we should 1st set **all** lanes of ucp_ep to failed_ep before destroying any of them?

Reply:
do you mean to make sure that we will not touch some UCT EP that will be destroyed from uct_ep_destroy() which may purge some outstanding operations?
If yes, I don't think if it is the problem, because the UCT EP which is going to be destroyed, still alive and can be used.

Reply:
is it possible purge of one lane will touch another lane?

Reply:
> is it possible purge of one lane will touch another lane?

I meant that purging outstanding WQE in uct_ep_destory() can lead that some GET completion on RC can be completed, and we could try sending RNDV ACK on already closed and set to `&ucp_failed_tl_ep` (or not closed yet and set to a real UCT EP). But either case is OK

Reply:
1. Is that completion with status UCS_OK or UCS_ERR_CANCELED?
2. Could we get RDNV GET completion on lane[0] and then try to send RNDV ACK on lane[1]?

Reply:
> * Is that completion with status UCS_OK or UCS_ERR_CANCELED?

https://github.com/openucx/ucx/blob/2869a9a386d02e9f203ce75d97e8813f401399e0/src/uct/ib/rc/accel/rc_mlx5_ep.c#L1035
it is with UCS_ERR_CANCELED
but I checked RNDV GET Zcopy, it doesn't send RNDV ACK in case of error:
https://github.com/openucx/ucx/blob/2869a9a386d02e9f203ce75d97e8813f401399e0/src/ucp/rndv/rndv.c#L581

> * Could we get RDNV GET completion on lane[0] and then try to send RNDV ACK on lane[1]?

I think it is possible in theory.

Reply:
> I think it is possible in theory.

So IMO it justifies setting all lanes to failed_ep and only then destroying them

Reply:
> > I think it is possible in theory.
> 
> So IMO it justifies setting all lanes to failed_ep and only then destroying them

fixed